clippy_lints/src/cognitive_complexity.rs

```diff
@@ -48,190 +52,258 @@ impl LintPass for CognitiveComplexity {
+    }
+
+    fn score_nd_structure(&mut self) {
+        self.add_to_score(self.current_nesting);
+        self.score_ni_structure();
+    }
+
+    fn score_ni_structure(&mut self) {
+        self.add_to_score(1);
+    }
+}
+
+// FIXME: delete these notes
+
+/*
+    Notes on debugging this (14/4):
+    // tests/ui/cognitive_complexity.rs
+    * Check why is 330 not being flagged (it has CoC 0 atm)
+*/
+
+impl<'ast> Visitor<'ast> for CoCHelper {
+    /*
+    # Implemented here:
+
+    ## Nesting Structures
+        IMPORTANT: ExprKind::Block(..)
+        already covers all cases.
+    ## (Nesting-Dependent) Increments
+        if
+        match
+        for, while, loop
+    ## (Nesting-Independent) Increments
+        break, continue
+        Sequences of binary logical operators
+        Recursive function calls
+        Macro calls
+    */
+
+    fn visit_expr(&mut self, ex: &'ast Expr) {
+        match ex.node {
+            // Nesting-Increasing (the one and only)
+            ExprKind::Block(..) => {
+                self.push_nesting();
+                walk_expr(self, ex);
+                self.pop_nesting();
+            },
+
+            ExprKind::Closure(.., _) | ExprKind::IfLet(..) => {
+                // If Let is free of own increment.
+                // This is because it's a language construct
+                // that's specifically designed to save on complexity.
+                walk_expr(self, ex);
+            },
+
+            // FIXME (FAR FUTURE): make a separate, documented case, for recursive calls,
+            // such that it's treated differently from function or method calls.
+            ExprKind::Call(..) | ExprKind::MethodCall(..) => {
+                self.score_ni_structure();
+                walk_expr(self, ex);
+            },
+
+            // Nesting-Dependent
+            ExprKind::If(..)
+            | ExprKind::Match(..)
+            | ExprKind::ForLoop(..)
+            | ExprKind::While(..)
+            | ExprKind::Loop(..) => {
+                // (For the IF-Case)
+                // Important: this pays for one "if" and one "else".
+                // Every "if" in an "else if" comes here again to pay
+                // for itself and its subsequent else.
+                self.score_nd_structure();
+                walk_expr(self, ex);
+            },
+
+            // Nesting-Independent
+            ExprKind::Mac(ref mac) => {
+                self.visit_mac(mac);
             },
-            ExprKind::Call(ref callee, _) => {
-                walk_expr(self, e);
-                let ty = self.cx.tables.node_type(callee.hir_id);
-                match ty.sty {
-                    ty::FnDef(..) | ty::FnPtr(_) => {
-                        let sig = ty.fn_sig(self.cx.tcx);
-                        if sig.skip_binder().output().sty == ty::Never {
-                            self.divergence += 1;
+
+            // Nesting-Independent
+            ExprKind::Break(_, _) | ExprKind::Continue(_) => {
```

Comment:
`break` can have an argument:

```rust
let x = loop {
    break 42;
};
```

Reply:
OHHHH. Wow, Rust is really an expression-oriented language. Things like these never cease to amaze me. Thanks!

Reply:
I've added this now ^^